In [1]:
%cd C:\\Users\\debie\\Documents\\anaconda_space

C:\Users\debie\Documents\anaconda_space


# Exploring eBay Car Sales Data

In this project, the dataset that will be explored contains data of used cars from the German eBay website. The aim of this project is to clean the data and analyze the included used car listings. 
The data dictionary provided with data is as follows:

    dateCrawled - When this ad was first crawled. All field-values are taken from this date.
    name - Name of the car.
    seller - Whether the seller is private or a dealer.
    offerType - The type of listing
    price - The price on the ad to sell the car.
    abtest - Whether the listing is included in an A/B test.
    vehicleType - The vehicle Type.
    yearOfRegistration - The year in which the car was first registered.
    gearbox - The transmission type.
    powerPS - The power of the car in PS.
    model - The car model name.
    kilometer - How many kilometers the car has driven.
    monthOfRegistration - The month in which the car was first registered.
    fuelType - What type of fuel the car uses.
    brand - The brand of the car.
    notRepairedDamage - If the car has a damage which is not yet repaired.
    dateCreated - The date on which the eBay listing was created.
    nrOfPictures - The number of pictures in the ad.
    postalCode - The postal code for the location of the vehicle.
    lastSeenOnline - When the crawler saw this ad last online.


In [3]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [4]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [5]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In this dataset, there are 371 527 rows and 20 columns. From the command .info() we can see that there are 5 columns (vehicleType / gearbox / model / fuelType / notRepairedDamage) which have null content that will need to be cleaned. We can see that some of the information are in German and will need to be translated in English. Then, in the powerPS column, some of the car are marked as having 0 PS which seems to mean that there is some information missing there.

The column names use camelcase instead on snakecase, which needs to be corrected.

In [6]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [7]:
new_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nb_pictures', 'postal_code',
       'last_seen']
autos.columns = new_names
print(autos.head(1))

          date_crawled        name  seller offer_type  price abtest  \
0  2016-03-24 11:52:17  Golf_3_1.6  privat    Angebot    480   test   

  vehicle_type  registration_year  gearbox  power_ps model  kilometer  \
0          NaN               1993  manuell         0  golf     150000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   

            ad_created  nb_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00            0        70435  2016-04-07 03:16:57  


We will look into the columns price and kilometer that are already numerical values.

In [11]:
autos.describe()

price  registration_year       power_ps      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nb_pictures   postal_code  
count       371528.000000     371528.0  371528.00000  
mean             5.734445          0.0   50820.66764  
std              3.712412          0.0   25799.08247  
min              0.000000          0.0    1067.00000  
25%              3.000000          0.0   30459.00000  
50%              6.000000          0.0   49610.00000  
75%              9.000000          0.0   71546.00000  
max             12.000000          0.0   99998.00000

In [24]:
print(autos["kilometer"].value_counts().sort_index())

5000        7069
10000       1949
20000       5676
30000       6041
40000       6376
50000       7615
60000       8669
70000       9773
80000      11053
90000      12523
100000     15920
125000     38067
150000    240797
Name: kilometer, dtype: int64


There doesn't seem that there is any issue with the kilometer columns. The kilometer count for each ad is already rounded.

In [41]:
print(autos["price"].value_counts(ascending=True))
print("Maximum price = ",autos["price"].max())
autos["price"].describe()

11449        1
6530         1
12973        1
19798        1
11987        1
         ...  
1200      4594
1000      4649
1500      5394
500       5670
0        10778
Name: price, Length: 5597, dtype: int64
Maximum price =  2147483647


count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

The prices values are a lot more diversified. We can see that the price that is the more observed is 0. It might mean that there are ad with the aim of getting rid off useless or old cars (which might be too expensive to repair for example).

The maximum price seems to be an error because it is around 2 billion euros which is not possible. Let's investigate and correct the high prices.

In [42]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

Après avoir regardé sur eBay, les voitures ne semblent pas dépasser le prix de 400 000€.

In [48]:
autos = autos[autos["price"]<400000]
autos["price"].describe()

count    371419.000000
mean       5733.580021
std        8922.004192
min           0.000000
25%        1150.000000
50%        2950.000000
75%        7200.000000
max      399997.000000
Name: price, dtype: float64

In [49]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

399997    1
395000    1
390000    1
370000    2
368000    1
350000    4
349000    1
345000    1
323223    1
300000    1
299000    3
295000    1
294900    1
285000    1
284000    1
279000    1
275000    1
270000    1
265000    1
260000    3
Name: price, dtype: int64

The price are now cleaned of absurd values.

In [54]:
date_crawled=autos["date_crawled"].str[:10]
ad_created=autos["ad_created"].str[:10]
last_seen=autos["last_seen"].str[:10]

In [61]:
print(date_crawled.value_counts(normalize=True,dropna=False).sort_index())

2016-03-05    0.025551
2016-03-06    0.014463
2016-03-07    0.035698
2016-03-08    0.033450
2016-03-09    0.034212
2016-03-10    0.032567
2016-03-11    0.032723
2016-03-12    0.036194
2016-03-13    0.015740
2016-03-14    0.036269
2016-03-15    0.033455
2016-03-16    0.030155
2016-03-17    0.031660
2016-03-18    0.013125
2016-03-19    0.035294
2016-03-20    0.036352
2016-03-21    0.035725
2016-03-22    0.032470
2016-03-23    0.031975
2016-03-24    0.029915
2016-03-25    0.032936
2016-03-26    0.031967
2016-03-27    0.030273
2016-03-28    0.035117
2016-03-29    0.034169
2016-03-30    0.033534
2016-03-31    0.031880
2016-04-01    0.034107
2016-04-02    0.035079
2016-04-03    0.038727
2016-04-04    0.037615
2016-04-05    0.012821
2016-04-06    0.003164
2016-04-07    0.001618
Name: date_crawled, dtype: float64


In [62]:
print(ad_created.value_counts(normalize=True,dropna=False).sort_index())

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.038886
2016-04-04    0.037744
2016-04-05    0.011653
2016-04-06    0.003155
2016-04-07    0.001556
Name: ad_created, Length: 114, dtype: float64


In [63]:
print(last_seen.value_counts(normalize=True,dropna=False).sort_index())

2016-03-05    0.001290
2016-03-06    0.004135
2016-03-07    0.005264
2016-03-08    0.008053
2016-03-09    0.009994
2016-03-10    0.011564
2016-03-11    0.013047
2016-03-12    0.023399
2016-03-13    0.008492
2016-03-14    0.012299
2016-03-15    0.016410
2016-03-16    0.016424
2016-03-17    0.028757
2016-03-18    0.006925
2016-03-19    0.016313
2016-03-20    0.019905
2016-03-21    0.020131
2016-03-22    0.020607
2016-03-23    0.018149
2016-03-24    0.019237
2016-03-25    0.019097
2016-03-26    0.016162
2016-03-27    0.016911
2016-03-28    0.022271
2016-03-29    0.023311
2016-03-30    0.023857
2016-03-31    0.024242
2016-04-01    0.024019
2016-04-02    0.025004
2016-04-03    0.025354
2016-04-04    0.025661
2016-04-05    0.126208
2016-04-06    0.217843
2016-04-07    0.129665
Name: last_seen, dtype: float64


In [64]:
autos["registration_year"].describe()

count    371419.000000
mean       2004.561043
std          91.936956
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

We can see here that there are some incorrect data as the minimum year of registration in 1000 and the maximum one is 9999.
Since the data is from 2016, let's consider the validate year of registration as between 1900 and 2016.

In [77]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True)

2000    0.068801
1999    0.063822
2005    0.062555
2006    0.056706
2001    0.056675
          ...   
1911    0.000003
1919    0.000003
1915    0.000003
1920    0.000003
1925    0.000003
Name: registration_year, Length: 97, dtype: float64

In [91]:
more_1=autos["brand"].value_counts(normalize=True)>0.01
brand=autos["brand"]
brand_top1=brand.value_counts(normalize=True)[more_1]
print(brand_top1)

volkswagen        0.212392
bmw               0.109698
opel              0.107090
mercedes_benz     0.095966
audi              0.089355
ford              0.068854
renault           0.047572
peugeot           0.029867
fiat              0.025763
seat              0.018636
skoda             0.015412
mazda             0.015350
smart             0.014108
citroen           0.013875
nissan            0.013572
toyota            0.012751
sonstige_autos    0.010589
Name: brand, dtype: float64


Above we have the top 1 % of the vehicle brand in the dataset.

Let's create a dictionnary which like them to their mean price.

In [107]:
dic = {}
for b in brand_top1.index:
    dic[b]=autos["price"][autos["brand"]==b].mean()
print(dic)

{'volkswagen': 5231.3883124769, 'bmw': 8234.726710455696, 'opel': 2870.0537738565854, 'mercedes_benz': 8398.008881357913, 'audi': 8849.650309058392, 'ford': 3595.8644488782115, 'renault': 2366.3506011315417, 'peugeot': 3206.2663099596357, 'fiat': 2803.857220589836, 'seat': 4420.369038664059, 'skoda': 6460.418046207022, 'mazda': 3970.6332420091326, 'smart': 3586.280802861685, 'citroen': 3649.0161648817943, 'nissan': 4585.39599256352, 'toyota': 5263.365875109938, 'sonstige_autos': 12652.675668519989}


Let's take interest into the 6 top brands : Volkswagen, BMW, Opel, Mercedes, Audi and Ford. We will look into their price and their mileage.

In [112]:
top6=brand_top1.head(6)
d_price={}
d_mileage={}
for b in top6.index:
    d_price[b]=autos["price"][autos["brand"]==b].mean()
    d_mileage[b]=autos["kilometer"][autos["brand"]==b].mean()
print(d_price)
print(d_mileage)

{'volkswagen': 5231.3883124769, 'bmw': 8234.726710455696, 'opel': 2870.0537738565854, 'mercedes_benz': 8398.008881357913, 'audi': 8849.650309058392, 'ford': 3595.8644488782115}
{'volkswagen': 128338.01943080415, 'bmw': 132663.37822986685, 'opel': 128756.83954237244, 'mercedes_benz': 130582.25481317012, 'audi': 129498.44686392018, 'ford': 123617.81831507798}


In [114]:
price_series=pd.Series(d_price)
mileage_series=pd.Series(d_mileage)
print(price_series)
print(mileage_series)

volkswagen       5231.388312
bmw              8234.726710
opel             2870.053774
mercedes_benz    8398.008881
audi             8849.650309
ford             3595.864449
dtype: float64
volkswagen       128338.019431
bmw              132663.378230
opel             128756.839542
mercedes_benz    130582.254813
audi             129498.446864
ford             123617.818315
dtype: float64


In [119]:
d={"mean_price":price_series,"mean_mileage":mileage_series}
top_brand=pd.DataFrame(data=d)
print(top_brand)

                mean_price   mean_mileage
volkswagen     5231.388312  128338.019431
bmw            8234.726710  132663.378230
opel           2870.053774  128756.839542
mercedes_benz  8398.008881  130582.254813
audi           8849.650309  129498.446864
ford           3595.864449  123617.818315


In [123]:
print(autos["seller"].value_counts())
print(autos["gearbox"].value_counts())
print(autos["unrepaired_damage"].value_counts())
print(autos["vehicle_type"].value_counts())

privat        356677
gewerblich         3
Name: seller, dtype: int64
manuell      263423
automatik     75025
Name: gearbox, dtype: int64
nein    255948
ja       35108
Name: unrepaired_damage, dtype: int64
limousine     95878
kleinwagen    80014
kombi         67550
bus           30196
cabrio        22891
coupe         18977
suv           14704
andere         3352
Name: vehicle_type, dtype: int64


We can see that in 4 columns, we need to replace german words wit english ones.

In [125]:
s_dic={"privat" : "private", "gewerblich" : "commercial"}
g_dic={"manuell" : "manual", "automatik" : "automatic"}
u_dic={"nein" : "no", "ja" : "yes"}
v_dic={"limousine" : "sedan",
    "kleinwagen" : "small car",
    "kombi" : "station wagon",
    "bus" : "bus",
    "cabrio" : "convertible",
    "coupe" : "coupe",
    "suv" : "suv",
    "andere" : "other"}

autos["seller"]=autos["seller"].replace(s_dic)
autos["gearbox"]=autos["gearbox"].replace(g_dic)
autos["unrepaired_damage"]=autos["unrepaired_damage"].replace(u_dic)
autos["vehicle_type"]=autos["vehicle_type"].replace(v_dic)

print(autos["seller"].value_counts())
print(autos["gearbox"].value_counts())
print(autos["unrepaired_damage"].value_counts())
print(autos["vehicle_type"].value_counts())

private       356677
commercial         3
Name: seller, dtype: int64
manual       263423
automatic     75025
Name: gearbox, dtype: int64
no     255948
yes     35108
Name: unrepaired_damage, dtype: int64
sedan            95878
small car        80014
station wagon    67550
bus              30196
convertible      22891
coupe            18977
suv              14704
other             3352
Name: vehicle_type, dtype: int64
